src/Workspaces/Core/Portable/FindSymbols/FindReferences/DependentTypeFinder.cs

```diff
@@ -25,573 +25,649 @@ namespace Microsoft.CodeAnalysis.FindSymbols
+        /// within the compilation.
         /// </summary>
-        private static readonly IEqualityComparer<SymbolKey> s_symbolIdComparer = SymbolKey.GetComparer(ignoreCase: true, ignoreAssemblyKeys: false);
+        private static readonly ConditionalWeakTable<Compilation, List<INamedTypeSymbol>> s_compilationAllAccessibleMetadataTypesTable =
+            new ConditionalWeakTable<Compilation, List<INamedTypeSymbol>>();
 
         /// <summary>
-        /// Used to create a new concurrent <see cref="SymbolKey"/> map for a given compilation when needed.
+        /// Used for implementing the Inherited-By relation for progression.
         /// </summary>
-        private static readonly ConditionalWeakTable<Compilation, ConcurrentDictionary<SymbolKey, List<SymbolKey>>>.CreateValueCallback s_createSymbolDictionary =
-            _ => new ConcurrentDictionary<SymbolKey, List<SymbolKey>>(s_symbolIdComparer);
+        internal static Task<IEnumerable<INamedTypeSymbol>> FindImmediatelyDerivedClassesAsync(
+            INamedTypeSymbol type,
+            Solution solution,
+            CancellationToken cancellationToken)
+        {
+            return FindDerivedClassesAsync(type, solution, projects: null,
+                transitive: false, cancellationToken: cancellationToken);
+        }
 
         /// <summary>
         /// This is an internal implementation of <see cref="SymbolFinder.FindDerivedClassesAsync"/>, which is a publically callable method.
         /// </summary>
-        public static Task<IEnumerable<INamedTypeSymbol>> FindDerivedClassesAsync(
+        public static Task<IEnumerable<INamedTypeSymbol>> FindTransitivelyDerivedClassesAsync(
             INamedTypeSymbol type,
             Solution solution,
             IImmutableSet<Project> projects,
             CancellationToken cancellationToken)
         {
-            // Only a class can have derived types.
-            if (type != null &&
-                type.TypeKind == TypeKind.Class &&
-                !type.IsSealed)
+            return FindDerivedClassesAsync(type, solution, projects,
+                transitive: true, cancellationToken: cancellationToken);
+        }
+
+        private static Task<IEnumerable<INamedTypeSymbol>> FindDerivedClassesAsync(
+            INamedTypeSymbol type,
+            Solution solution,
+            IImmutableSet<Project> projects, 
+            bool transitive,
+            CancellationToken cancellationToken)
+        {
+            if (s_isNonSealedClass(type))
             {
-                return GetDependentTypesAsync(
-                    type,
-                    solution,
-                    projects,
-                    s_findDerivedClassesPredicate,
-                    s_derivedClassesCache,
-                    cancellationToken);
+                Func<HashSet<INamedTypeSymbol>, INamedTypeSymbol, bool> metadataTypeMatches =
+                    (s, t) => TypeDerivesFrom(s, t, transitive);
```

Comment:
Nit: variable names longer than `s` and `t`? Normally wouldn't care, but can't figure out from context what they are.